core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -295,21 +339,23 @@ private boolean pollForCompletion(BlockingQueue<CompletableFuture<?>> queue, lon
          expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, entry.getCreated(), currentTime);
       }
       if (expiredMortal) {
-         CompletableFuture<Void> future = handleLifespanExpireEntry(entry.getKey(), value, lifespan, hasLock);
-         if (waitOnLifespanExpiration(hasLock)) {
-            return future.thenApply(ignored -> true);
+         CompletableFuture<Boolean> future = handleLifespanExpireEntry(entry.getKey(), value, lifespan, isWrite);
+         if (waitOnLifespanExpiration(isWrite)) {
+            return future;
          }
          // We don't want to block the user while the remove expired is happening for lifespan
```

Comment:
Added for "on a read" to clarify that.

Reply:
IDK, there's nothing that stops `waitOnLifeSpanExpiration(false)` from blocking.

Reply:
well it is also not a statement of fact, but rather what we want to do :) But currently if the read finds a lifespan expired entry it doesn't block, except for when optimistic tx are enabled, which I don't want to block, but have to, haha.

Reply:
Ok, but you don't even get to this line if `waitOnLifespanExpiration()` decided for you that you should block :)

Perhaps it would make more sense if the comment was above the `if`?

Reply:
I moved the line up.